File: weather.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <float.h>
#include <time.h>
#include <string.h>
#include "csv.h"

char presentCSV[20][50] = {
  "./data/2021_tor_data.csv",
  "./data/2021_van_data.csv",
  "./data/2021_mtl_data.csv",
  "./data/2021_cal_data.csv",
  "./data/2021_edm_data.csv",
  "./data/2021_ham_data.csv",
  "./data/2021_hfx_data.csv",
  "./data/2021_kel_data.csv",
  "./data/2021_kit_data.csv",
  "./data/2021_lon_data.csv",
  "./data/2021_osh_data.csv",
  "./data/2021_ott_data.csv",
  "./data/2021_peg_data.csv",
  "./data/2021_qbc_data.csv",
  "./data/2021_reg_data.csv",
  "./data/2021_sas_data.csv",
  "./data/2021_stcath_data.csv",
  "./data/2021_stjn_data.csv",
  "./data/2021_vic_data.csv",
  "./data/2021_win_data.csv"
};

char prevCSV[20][50] = {
  "./data/2020_tor_data.csv",
  "./data/2020_van_data.csv",
  "./data/2020_mtl_data.csv",
  "./data/2020_cal_data.csv",
  "./data/2020_edm_data.csv",
  "./data/2020_ham_data.csv",
  "./data/2020_hfx_data.csv",
  "./data/2020_kel_data.csv",
  "./data/2020_kit_data.csv",
  "./data/2020_lon_data.csv",
  "./data/2020_osh_data.csv",
  "./data/2020_ott_data.csv",
  "./data/2020_peg_data.csv",
  "./data/2020_qbc_data.csv",
  "./data/2020_reg_data.csv",
  "./data/2020_sas_data.csv",
  "./data/2020_stcath_data.csv",
  "./data/2020_stjn_data.csv",
  "./data/2020_vic_data.csv",
  "./data/2020_win_data.csv" 
};

// Calculate the euclidian distance of two arrays
double calcEuclidianDistance(
  double *presentMeans, 
  double *prevMeans, 
  int numPresentMeans
) {
  double euclidianDistance = 0;
  for (int i = 0; i < numPresentMeans; i++) {
    euclidianDistance += pow(presentMeans[i] - prevMeans[i], 2);
  }
  return sqrt(euclidianDistance);
}

// Determine the minimum value from an array
double min(double *arr, int len) {
  double minVal = DBL_MAX;
  for (int i = 0; i < len; i++)
    if (arr[i] < minVal)
      minVal = arr[i];
  return minVal;
}

// Predicts the weather of a particular day of a year
void predictWeatherForGivenDay(
  double *prediction, 
  double *actual,
  char *presentCSV,
  char *prevCSV, 
  int day, 
  int *cols, 
  int numParams
) {
  // Get current data for the past 7 days
  char CD[7][50][50];
  readCSV(CD, presentCSV, day - 7, day - 1, cols, numParams);

  // Get previous data (1 year old) for a 14 day period
  char PD[14][50][50];
  readCSV(PD, prevCSV, day - 7, day + 6, cols, numParams);

  // 8 sliding windows based on previous data, 7 days per window
  double windows[8][7][numParams];
  memset(windows, 0, 8*7*numParams*sizeof(double));
  for (int i = 0; i < 8; i++) {
    for (int j = 0; j < 7; j++) {
      for (int k = 0; k < numParams; k++) {
        windows[i][j][k] = atof(PD[j+i][k]);
      }
    }
  }

  // Sum the present values 
  double presentSums[numParams];
  memset(presentSums, 0, numParams*sizeof(double));
  for (int i = 0; i < 7; i++)
    for (int j = 0; j < numParams; j++)
      presentSums[j] += atof(CD[i][j]);

  // Using the present sums, determine the means
  double presentMeans[numParams];
  memset(presentMeans, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    presentMeans[i] = presentSums[i] / 7;

  // Determine the means of the previous values based on the
  // sliding windows and calculate the Euclidian Distances
  double edList[8];
  memset(edList, 0, 8*sizeof(double));
  for (int i = 0; i < 8; i++) {
    double prevSums[numParams];
    memset(prevSums, 0, numParams*sizeof(double));
    for (int j = 0; j < 7; j++)
      for (int k = 0; k < numParams; k++)
        prevSums[k] += windows[i][j][k];

    double prevMeans[numParams];
    memset(prevMeans, 0, numParams*sizeof(double));
    for (int j = 0; j < numParams; j++)
      prevMeans[j] = prevSums[j] / 7;

    int numPresentMeans = sizeof(presentMeans) / sizeof(presentMeans[0]);
    double euclidDistance = calcEuclidianDistance
    (
      presentMeans,
      prevMeans,
      numPresentMeans
    );

    edList[i] = euclidDistance;
  }

  // Determine the minimum Euclidian Distance and find its index
  double minEd = min(edList, sizeof(edList) / sizeof(edList[0]));
  int indexMinEd = 0;
  for (int i = 0; i < 8; i++) {
    if (edList[i] == minEd) {
      indexMinEd = i;
      break;
    }
  }

  // Isolate the window with the minimum Euclidian Distance
  double selectedPrevWindow[7][numParams];
  memset(selectedPrevWindow, 0, 7*numParams*sizeof(double));
  for (int i = 0; i < 7; i++)
    for (int j = 0; j < numParams; j++)
      selectedPrevWindow[i][j] = windows[indexMinEd][i][j];

  // Find variation vector based on present data
  double presentVariationVector[numParams][6];
  memset(presentVariationVector, 0, numParams*6*sizeof(double));
  for (int i = 0; i < numParams; i++)
    for (int j = 0; j < 6; j++)
      presentVariationVector[i][j] = atof(CD[j+1][i]) - atof(CD[j][i]);

  // Find variation vector based on previous data
  double prevVariationVector[numParams][6];
  memset(prevVariationVector, 0, numParams*6*sizeof(double));
  for (int i = 0; i < numParams; i++)
    for (int j = 0; j < 6; j++)
      prevVariationVector[i][j] = selectedPrevWindow[j+1][i] - selectedPrevWindow[j][i];

  // Determine the mean of the present variation vector
  double meanPresentVar[numParams];
  memset(meanPresentVar, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++) {
    double presentVarSum = 0;
    for (int j = 0; j < 6; j++)
      presentVarSum += presentVariationVector[i][j];
    meanPresentVar[i] = presentVarSum / 6;
  }

  // Determine the mean of the previous variation vector
  double meanPrevVar[numParams];
  memset(meanPrevVar, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++) {
    double prevVarSum = 0;
    for (int j = 0; j < 6; j++)
      prevVarSum += prevVariationVector[i][j];
    meanPrevVar[i] = prevVarSum / 6;
  }

  // Determine the mean of both present and previous mean variation vectors
  double meanVariationVector[numParams];
  memset(meanVariationVector, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    meanVariationVector[i] = (meanPresentVar[i] + meanPrevVar[i]) / 2;

  // Retrieve values from the day previous to the day being predicted
  double previousDay[numParams];
  memset(previousDay, 0, numParams*sizeof(double));
  for (int i = 0; i < numParams; i++)
    previousDay[i] = atof(CD[6][i]);

  // Add mean variation vector values to the previous day to get final prediction
  for (int i = 0; i < numParams; i++)
    previousDay[i] += meanVariationVector[i];

  // Assign the final prediction values
  for (int i = 0; i < numParams; i++)
    prediction[i] = previousDay[i];

  // Retrive the actual values for comparision purposes
  char actualValues[1][50][50];
  readCSV(actualValues, presentCSV, day, day, cols, numParams);
  for (int i = 0; i < numParams; i++)
    actual[i] = atof(actualValues[0][i]);
}

int main() {
  clock_t t;
  t = clock();

  // 9 - Max Temp (°C)
  // 11 - Min Temp (°C)
  // 13 - Mean Temp (°C)
  // 15 - Heat Deg Days (°C)
  // 17 - Cool Deg Days (°C)
  // 19 - Total Rain (mm)
  // 21 - Total Snow (cm)
  // 23 - Total Precip (mm)
  // 25 - Snow on Grnd (cm)
  int cols[] = {9, 11, 13, 15, 17, 19, 21, 23, 25};
  int numParams = sizeof(cols) / sizeof(cols[0]);
  int numLocations = sizeof(presentCSV) / sizeof(presentCSV[0]);

  for (int location = 0; location < numLocations; location++) {
    for (int i = 8; i <= 359; i++) {
      double prediction[numParams];
      double actual[numParams];
      predictWeatherForGivenDay(
        prediction, 
        actual, 
        presentCSV[location], 
        prevCSV[location], 
        i,
        cols, 
        numParams
      );
      
      // Comment out these print statements for a more accurate time measurement
      // printf("\nDay %d\nActual: ", i);
      // for (int j = 0; j < numParams; j++) {
      //   printf("%.2lf ", actual[j]);
      // }
      // printf("\nPrediction: ");
      // for (int j = 0; j < numParams; j++) {
      //   printf("%.2lf ", prediction[j]);
      // }
      // printf("\n");
    }
  }
  
  t = clock() - t;
  double timeTaken = ((double)t)/CLOCKS_PER_SEC;
  printf("\nTime taken: %f seconds\n", timeTaken);

  return 0;
}

File: weather.py
from operator import indexOf
import pandas as pd
import math
from pprint import pprint
import time
import matplotlib.pyplot as plt
import numpy as np

data_files_2020 = [
    pd.read_csv("data/2020_data.csv"),
    pd.read_csv("data/2020_van_data.csv"),
    pd.read_csv("data/2020_mtl_data.csv"),
    pd.read_csv("data/2020_cal_data.csv"),
    pd.read_csv("data/2020_edm_data.csv"),
    pd.read_csv("data/2020_ham_data.csv"),
    pd.read_csv("data/2020_hfx_data.csv"),
    pd.read_csv("data/2020_kel_data.csv"),
    pd.read_csv("data/2020_kit_data.csv"),
    pd.read_csv("data/2020_lon_data.csv"),
    pd.read_csv("data/2020_osh_data.csv"),
    pd.read_csv("data/2020_ott_data.csv"),
    pd.read_csv("data/2020_peg_data.csv"),
    pd.read_csv("data/2020_qbc_data.csv"),
    pd.read_csv("data/2020_reg_data.csv"),
    pd.read_csv("data/2020_sas_data.csv"),
    pd.read_csv("data/2020_stcath_data.csv"),
    pd.read_csv("data/2020_stjn_data.csv"),
    pd.read_csv("data/2020_vic_data.csv"),
    pd.read_csv("data/2020_win_data.csv"),
    ]

data_files_2021 = [
    pd.read_csv("data/2021_data.csv"),
    pd.read_csv("data/2021_van_data.csv"),
    pd.read_csv("data/2021_mtl_data.csv"),
    pd.read_csv("data/2021_cal_data.csv"),
    pd.read_csv("data/2021_edm_data.csv"),
    pd.read_csv("data/2021_ham_data.csv"),
    pd.read_csv("data/2021_hfx_data.csv"),
    pd.read_csv("data/2021_kel_data.csv"),
    pd.read_csv("data/2021_kit_data.csv"),
    pd.read_csv("data/2021_lon_data.csv"),
    pd.read_csv("data/2021_osh_data.csv"),
    pd.read_csv("data/2021_ott_data.csv"),
    pd.read_csv("data/2021_peg_data.csv"),
    pd.read_csv("data/2021_qbc_data.csv"),
    pd.read_csv("data/2021_reg_data.csv"),
    pd.read_csv("data/2021_sas_data.csv"),
    pd.read_csv("data/2021_stcath_data.csv"),
    pd.read_csv("data/2021_stjn_data.csv"),
    pd.read_csv("data/2021_vic_data.csv"),
    pd.read_csv("data/2021_win_data.csv")
]

PARAMS = [
    "Max Temp (°C)",
    "Min Temp (°C)",
    "Mean Temp (°C)",
    "Heat Deg Days (°C)",
    "Cool Deg Days (°C)",
    "Total Rain (mm)",
    "Total Snow (cm)",
    "Total Precip (mm)",
    "Snow on Grnd (cm)",
]

NUM_OF_PARAMS = len(PARAMS)

def get_mean(arr):
    return [val / 7 for val in arr]

# Predicting any day in 2021
def forecast_day(day,data_21,data_20):
    if day < 7 or day > 360:
        return None, None

    present_days = data_21.loc[day-7:day-1]
    present_days = present_days[PARAMS]
    present_days = present_days.fillna(0)
    CD = present_days.to_numpy()

    prev_days = data_20.loc[day-7:day+6]
    prev_days = prev_days[PARAMS]

    prev_days = prev_days.fillna(0)
    PD = prev_days.to_numpy()

    windows = []
    for i in range(len(PD) - 6):
        windows.append(PD[i: i+7])

    # calculating means for present year 7 days
    present_sums = [0] * NUM_OF_PARAMS
    for line in CD:
        for i in range(len(line)):
            if not math.isnan(line[i]): present_sums[i] += line[i] 
    present_means = get_mean(present_sums)

    # mean of parameters for each window
    ed_list = []
    for window in windows:
        prev_sums = [0] * NUM_OF_PARAMS
        for vals in window:
            for i in range(len(prev_sums)):
                prev_sums[i] += vals[i]
        prev_means = get_mean(prev_sums)

        euclid_distance = sum([(present_means[i] - prev_means[i]) ** 2 for i in range(len(present_means))])
        ed_list.append(euclid_distance)

    min_ed = min(ed_list)
    index_min_ed = indexOf(ed_list, min_ed)
    selected_prev_window = windows[index_min_ed]

    present_variation_vector = []
    for i in range(len(CD[0])):
        vector = []
        for j in range(len(CD)-1):
            vector.append(CD[j+1][i] - CD[j][i])
        present_variation_vector.append(vector)

    prev_variation_vector = []
    for i in range(len(selected_prev_window[0])):
            vector = []
            for j in range(len(selected_prev_window)-1):
                vector.append(selected_prev_window[j+1][i] - selected_prev_window[j][i])
            prev_variation_vector.append(vector)
    
    mean_present_var = []
    for i in range(len(present_variation_vector)):
        mean_present_var.append(sum(present_variation_vector[i]) / len(present_variation_vector[i]))

    mean_prev_var = []
    for i in range(len(prev_variation_vector)):
        mean_prev_var.append(sum(prev_variation_vector[i]) / len(prev_variation_vector[i]))

    mean_variation_vector = []
    for i in range(len(mean_present_var)):
        mean_variation_vector.append((mean_present_var[i] + mean_prev_var[i]) / 2)
    
    prev_day = CD[-1]
    for i in range(len(prev_day)):
        prev_day[i] += mean_variation_vector[i]
    pred_arr = prev_day.tolist()
    
    for i in range(len(pred_arr[5:])):
        if pred_arr[5+i] < 0:
            pred_arr[5+i] = 0

    actual_data = data_21.loc[day]
    actual_data = actual_data.fillna(0)
    actual_values = actual_data[PARAMS].to_numpy().flatten()

    return actual_values, pred_arr

if __name__ == "__main__":
    t0 = time.time()
    day_range = np.arange(0, 365)
    actual_list = [[] for i in range(20)]
    pred_list = [[] for i in range(20)]


    for day in day_range:
        for i in range(len(pred_list)):
            temp_actual, temp_pred = forecast_day(day,data_files_2021[i],data_files_2020[i])

            if temp_pred:
                actual_list[i].append(list(temp_actual))
                pred_list[i].append(list(temp_pred))

            else:
                actual_list[i].append([0] * NUM_OF_PARAMS)
                pred_list[i].append([0] * NUM_OF_PARAMS)

    print(f"\nTotal time: {time.time() - t0}")

    toronto_actual_data = np.transpose(actual_list[0])
    toronto_pred_data = np.transpose(pred_list[0])


    plt.figure()
    plt.subplot(1, 3, 1)
    plt.gcf().set_size_inches(20,12)

    index = 0
    for param in PARAMS:
        plt.subplot(2, 5, index+1)
        plt.plot(day_range, toronto_actual_data[index], label='Toronto Actual', color = "red")
        plt.plot(day_range, toronto_pred_data[index], label='Toronto Prediction', color = "blue")
        plt.ylim(-20, 40)
        plt.title(f"{param}")
        plt.legend(['Toronto Actual', 'Toronto Prediction'])
        index += 1
    plt.show()
